### ISSUE TITLE
Azure AD B2C Profile Edit Policy pass through

### ISSUE THREAD

**Issue by cletisgipson** [IdentityServer/IdentityServer3#2816]
### Question / Issue

I am using IdentityServer3 as a pass through to Azure AD B2C as part of a bigger SSO project.  I have managed to get this all working. The only thing, that I have not got working is the profile edit request?

With Azure AD B2C they require a policy id, this works fine for the sign in/sign up, but not the edit profile policy.  Is there a way to intercept these requests coming into IdentityServer3, check for these policies and if one is found, pass it on through to azure ad b2c or do the normal login flow?  

Hopefully, this makes sense.

**Comment by brockallen** [comment]
You must have plugged in the OIDC middleware to connect to Azure, yes? You are the one that controls that, which also means you can add any custom params they need.

**Comment by cletisgipson** [comment]
Yes, I did plugin in the OIDC middleware. 

This is the path my app takes..  custom app -> identity server -> azure b2c.  

When I issue a call from the custom app to identity server for the profile edit, it just returns me back to my app instead of going on to microsoft to edit the profile.    Here is my identity provider configuration in identityserver.

`            var openid = new Microsoft.Owin.Security.OpenIdConnect.OpenIdConnectAuthenticationOptions
            {
                AuthenticationType = "Microsoft",
                Caption = "Indiana One Access",
                SignInAsAuthenticationType = signInAsType,
                ClientId = ConfigurationManager.AppSettings["ida.microsoft.clientid"],
                ClientSecret = ConfigurationManager.AppSettings["ida.microsoft.clientsecret"],
                RedirectUri = ConfigurationManager.AppSettings["ida.microsoft.redirecturi"],
                PostLogoutRedirectUri = ConfigurationManager.AppSettings["ida.microsoft.postlogoutredirecturi"],
                MetadataAddress = ConfigurationManager.AppSettings["ida.microsoft.metadataaddress"],
                ConfigurationManager = new PolicyConfigurationManager(ConfigurationManager.AppSettings["ida.microsoft.metadataaddress"], new string[] { ConfigurationManager.AppSettings["ida.microsoft.signinorsignuppolicy"], ConfigurationManager.AppSettings["ida.microsoft.editprofilepolicy"] })

            };

            app.UseOpenIdConnectAuthentication(openid);`

The call from my custom app....

`    public class ProfileController : Controller
    {
        // GET: Profile
        public void Index()
        {
            if (User.Identity.IsAuthenticated)
            {
                HttpContext.GetOwinContext().Authentication.Challenge(
                    new AuthenticationProperties(
                        new Dictionary<string, string>
                        {
                            {"doeb2c", "B2C_1_ProfilePolicy"}
                        })
                    {
                        RedirectUri = "/",
                    }, OpenIdConnectAuthenticationDefaults.AuthenticationType);
            }
        }
    }`

Should the profile edit request come from Identity Server itself or my custom app?  If it is to come from Identity Server, I will need to after the profile edit, return the user back through to my custom app.

**Comment by brockallen** [comment]
I don't know -- I do not know how Azure B2C works.

**Comment by cletisgipson** [comment]
Okay one last question.

With IdentityServer, if I have a controller created with an action, is there a way to issue a challenge to my external identity provider?  This might solve the problem, if I can only issue a challenge to the external provider.

**Comment by brockallen** [comment]
If you pass the idp:yourIdp in the acr_values to IdSvr from your client, then we will automatically trigger that one, rather than prompting the user. That's covered in our docs: https://identityserver.github.io/Documentation/docsv2/endpoints/authorization.html

**Comment by cletisgipson** [comment]
My implementation of IdSrv already redirects automatically to the provider.  With Azure B2C, they use a thing called policies.  Say I have a SIGNINORSIGNUP policy and a PROFILEEDIT policy.  From my relying party, when it is issues a login request because of authorization, it adds the SIGNINORSIGNUP policy as a parameter.  If I am not already logged into IDSrv, I will get forwarded to Azure B2C, this is the expected behavior.  After I am logged in and returned back to my relying party. I issue a request using the PROFILEEDIT policy, it looks the exact same as signin request, except the policy parameter is changed. When I get to IDSrv with the new request, IDSrv alread sees that I am logged in and sends me back to my relying party.  I would like to figure out a way, for IDSrv, to look at the policy parameter and if it sees the PROFILEEDIT value, it will pass that request onto Azure B2C regardless of the logged in status for the user in IDSrv.  I hope this makes it a little more clear on what I am trying to do.

Here is a link to the quickstart for .net mvc web app, provided by microsoft.

[Azure B2C Quickstart](https://azure.microsoft.com/en-us/documentation/articles/active-directory-b2c-devquickstarts-web-dotnet/)

And here is the bit of code that is fired, when doing the basic sign in/sign out and profile edit actions.

[Quickstart Account Controller code.](https://github.com/AzureADQuickStarts/B2C-WebApp-OpenIdConnect-DotNet/blob/complete/WebApp-B2C-DotNet/Controllers/AccountController.cs)

**Comment by brockallen** [comment]
If you want to modify the request params to Azure from IdSvr, then hook the notification events on the OIDC  middleware.

**Comment by cletisgipson** [comment]
One more question...I have figured out how to update the acr_values that are getting passed from my RP to Identity Server.  I am passing a "profile:true" or "password:true", which I know how to access in IdentityServer notifications and my relying party notifications.  IdentityServer only redirects to the external identity provider, when the user is not logged in.  When the user is already logged in IDSrv automatically redirects me back to my RP.  What I need to do is intercept the authorize request and evaluate the acr_values.  If I find the values that I passed,  I want to redirect to the external provider regardless of the logged in status of the user.
